apps/dash-tsne/demo.py

```diff
@@ -0,0 +1,624 @@
+                                        },
+                                    ),
+                                    html.Div(id="div-plot-click-image"),
+                                    html.Div(id="div-plot-click-wordemb"),
+                                ],
+                            )
+                        ],
+                    ),
+                ],
+            ),
+        ],
+    )
+
+
+def demo_callbacks(app):
+    def generate_figure_image(groups, layout):
+        data = []
+
+        for idx, val in groups:
+            scatter = go.Scatter3d(
+                name=idx,
+                x=val["x"],
+                y=val["y"],
+                z=val["z"],
+                text=[idx for _ in range(val["x"].shape[0])],
+                textposition="top center",
+                mode="markers",
+                marker=dict(size=3, symbol="circle"),
+            )
+            data.append(scatter)
+
+        figure = go.Figure(data=data, layout=layout)
+
+        return figure
+
+    # Scatter Plot of the t-SNE datasets
+    def generate_figure_word_vec(
+        embedding_df, layout, wordemb_display_mode, selected_word, dataset
+    ):
+
+        try:
+            # Regular displays the full scatter plot with only circles
+            if wordemb_display_mode == "regular":
+                plot_mode = "markers"
+
+            # Nearest Neighbors displays only the 200 nearest neighbors of the selected_word, in text rather than circles
+            elif wordemb_display_mode == "neighbors":
+                if not selected_word:
+                    return go.Figure()
+
+                plot_mode = "text"
+
+                # Get the nearest neighbors indices using Euclidean distance
+                vector = data_dict[dataset].set_index("0")
+                selected_vec = vector.loc[selected_word]
+
+                def compare_pd(vector):
+                    return spatial_distance.euclidean(vector, selected_vec)
+
+                distance_map = vector.apply(compare_pd, axis=1)
+                neighbors_idx = distance_map.sort_values()[:100].index
+
+                # Select those neighbors from the embedding_df
+                embedding_df = embedding_df.loc[neighbors_idx]
+
+            scatter = go.Scatter3d(
+                name=str(embedding_df.index),
+                x=embedding_df["x"],
+                y=embedding_df["y"],
+                z=embedding_df["z"],
+                text=embedding_df.index,
+                textposition="middle center",
+                showlegend=False,
+                mode=plot_mode,
+                marker=dict(size=3, color="#3266c1", symbol="circle"),
+            )
+
+            figure = go.Figure(data=[scatter], layout=layout)
+
+            return figure
+        except KeyError as error:
+            print(error)
+            raise PreventUpdate
+
+    @app.server.before_first_request
+    def load_image_data():
+        global data_dict
```

Comment:
Is a `global` variable necessary in this case?

Reply:
If it is, feel free to ignore this comment. However, I think for Dash apps we are supposed to avoid using `global` variables and if we need them we should use `dcc.Store`. I think you can just declare `data_dict` without using `global`.